fortuna/data/loader/base.py

```diff
@@ -0,0 +1,379 @@
+from __future__ import annotations
+
+from typing import Iterable, Callable, Optional, Tuple, List, TypeVar, Type
+import abc
+
+import jax
+from flax import jax_utils
+from jax.tree_util import tree_map
+
+from fortuna.data.loader.utils import IterableData
+from fortuna.typing import Batch, InputData, Array, Status, Targets
+
+T = TypeVar('T')
+
+
+class BaseDataLoaderABC(abc.ABC):
+    def __init__(self, iterable: IterableData):
+        self._iterable = iterable
+
+    def __iter__(self) -> Iterable[Batch]:
+        yield from self._iterable
+
+    @property
+    def size(self) -> int:
+        """
+        The number of data points in the data loader.
+
+        Returns
+        -------
+        int
+            Number of data points.
+        """
+        c = 0
+        for inputs, targets in self:
+            if isinstance(inputs, dict):
+                inputs = inputs[list(inputs.keys())[0]]
+            c += inputs.shape[0]
+        return c
+
+    @abc.abstractmethod
+    def to_inputs_loader(self) -> BaseInputsLoader:
+        """
+        Reduce a data loader to an inputs loader.
+
+        Returns
+        -------
+        BaseInputsLoader
+            The inputs loader derived from the data loader. This will be a concrete instance of a subclass 
+            of :class:`~fortuna.data.loader.BaseInputsLoader`.
+        """
+        pass
+
+    @abc.abstractmethod
+    def to_targets_loader(self) -> BaseTargetsLoader:
+        """
+        Reduce a data loader to a targets loader.
+
+        Returns
+        -------
+        BaseTargetsLoader
+            The targets loader derived from the data loader. This will be a concrete instance of a subclass 
+            of :class:`~fortuna.data.loader.BaseTargetsLoader`.
+        """
+        pass
+
+    @classmethod
+    def from_callable_iterable(
+        cls: Type[T],
+        fun: Callable[
+            [],
+            Iterable[Batch],
+        ],
+    ) -> T:
+        """
+        Transform a callable iterable into a concrete instance of a subclass of 
+        :class:`~fortuna.data.loader.BaseDataLoader`.
+
+        Parameters
+        ----------
+        fun: Callable[[], Iterable[Batch]]
+            A callable iterable of tuples of input and target arrays.
```

Comment:
Non necessarily arrays any more?

Reply:
That's right. We should allow fortuna to work with other types of input data (e.g., when using the `transfromers` library the input data is a `Dict[str, jax.Array]`)